receiver/iisreceiver/scraper.go

```diff
@@ -0,0 +1,333 @@
+
+//go:build windows
+// +build windows
+
+package iisreceiver // import "github.com/open-telemetry/opentelemetry-collector-contrib/receiver/iisreceiver"
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/consumer"
+	"go.opentelemetry.io/collector/model/pdata"
+	"go.opentelemetry.io/collector/receiver/scraperhelper"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/receiver/iisreceiver/internal/metadata"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/receiver/windowsperfcountersreceiver"
+)
+
+type iisReceiver struct {
+	params              component.ReceiverCreateSettings
+	config              *Config
+	consumer            consumer.Metrics
+	perfCounterReceiver component.MetricsReceiver
+}
+
+// new returns a iisReceiver
+func newIisReceiver(params component.ReceiverCreateSettings, cfg *Config, consumer consumer.Metrics) *iisReceiver {
+	return &iisReceiver{params: params, config: cfg, consumer: consumer}
+}
+
+// Start creates and starts the prometheus receiver.
+func (rcvr iisReceiver) Start(ctx context.Context, host component.Host) error {
+	wFactory := windowsperfcountersreceiver.NewFactory()
+
+	wConfig, err := getWindowsPerfConfig(rcvr.config.ScraperControllerSettings, metadata.NewMetricsBuilder(rcvr.config.metricSettings), getPerfCounters())
+	if err != nil {
+		return fmt.Errorf("failed to create windows performance counter receiver config: %v", err)
+	}
+
+	wpcRcvr, err := wFactory.CreateMetricsReceiver(ctx, rcvr.params, wConfig, rcvr.consumer)
+	if err != nil {
+		return fmt.Errorf("failed to create windows performance counter receiver: %v", err)
+	}
+
+	rcvr.perfCounterReceiver = wpcRcvr
+	return rcvr.perfCounterReceiver.Start(ctx, host)
+}
+
+func generateVals(metricBuilder *metadata.MetricsBuilder) {
+	now := pdata.NewTimestampFromTime(time.Now())
+	metricBuilder.RecordIisConnectionActiveDataPoint(now, 0)
+	metricBuilder.RecordIisConnectionAnonymousDataPoint(now, 0)
+	metricBuilder.RecordIisConnectionAttemptCountDataPoint(now, 0)
+	metricBuilder.RecordIisNetworkBlockedDataPoint(now, 0)
+	metricBuilder.RecordIisNetworkFileCountDataPoint(now, 0, "")
+	metricBuilder.RecordIisNetworkIoDataPoint(now, 0, "")
+	metricBuilder.RecordIisRequestCountDataPoint(now, 0, "")
+	metricBuilder.RecordIisRequestQueueAgeMaxDataPoint(now, 0)
+	metricBuilder.RecordIisRequestQueueCountDataPoint(now, 0)
+	metricBuilder.RecordIisRequestRejectedDataPoint(now, 0)
+	metricBuilder.RecordIisThreadActiveDataPoint(now, 0)
+	metricBuilder.RecordIisUptimeDataPoint(now, 0)
+}
+
+func getWindowsPerfConfig(scraperCfg scraperhelper.ScraperControllerSettings, metricBuilder *metadata.MetricsBuilder, perfCounters map[string][]windowsperfcountersreceiver.PerfCounterConfig) (*windowsperfcountersreceiver.Config, error) {
+	generateVals(metricBuilder)
+	metricSlice := metricBuilder.Emit().ResourceMetrics().At(0).InstrumentationLibraryMetrics().At(0).Metrics()
```

Comment:
Not sure what's going on here but it doesn't look right.

Reply:
So, there's no static representation of the metrics to be pointed at based on whether or not they are enabled. Emit only returns metrics that actually have a datapoint attached to it. So, we are generating atleast one datapoint for every metric, it only adds the datapoint based on whether or not the metric is enabled in the `MetricSettings`. So, then we have a slice of metrics that we can loop through to build the config for based on the representation of the metrics. We could have a static map that defines each individual metric config but then if metadata changes we have to remember to update the static map as well.